### PR TITLE
[FLINK-4900] flink-master: Allow to deploy TM with container

### PR DESCRIPTION
**Pull request by Makman2**
Allows via a setting to deploy a base image on that a task manager runs.

### CODE REVIEW

**Comment by Makman2** [comment]
Hello :)
Not sure what has gone wrong in the tests (it seems the CI is not really stable when looking at other PRs, especially on my typo fixing PR), though it would be good to merge it asap, as I'm working on getting flink into DC/OS and having this feature inside master would be a step forward as it makes building docker images easier :) thx in advance :+1:
CC @EronWright

**Comment by mxm** [comment]
Thanks for the PR! Jenkins CI is currently broken but Travis CI passed. Looks good to me but would be great if you could take a look @EronWright.

File: flink-core/src/main/java/org/apache/flink/configuration/ConfigConstants.java [hunk 1]
```diff
@@ -489,6 +489,29 @@
 	 */
 	public static final String MESOS_RESOURCEMANAGER_TASKS_CPUS = "mesos.resourcemanager.tasks.cpus";
 
+	/**
+	 * The base container image to use for task managers (started via the unified containerizer).
```
Comment (EronWright) [inline flink-core/src/main/java/org/apache/flink/configuration/ConfigConstants.java]: Please rephrase the comment: "The container image to use for task managers."    In other words, it isn't a 'base' image in this context, and numerous containerizers are now supported.

File: flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/MesosApplicationMasterRunner.java [hunk 2]
```diff
@@ -601,6 +601,54 @@ else if (recoveryMode == HighAvailabilityMode.ZOOKEEPER) {
...
+								.setType(Protos.Image.Type.DOCKER)
+								.setDocker(Protos.Image.Docker.newBuilder()
+									.setName(taskManagerContainerName))));
+					break;
+				case ConfigConstants.MESOS_RESOURCEMANAGER_TASKS_CONTAINER_TYPE_DOCKER:
+					containerInfo = Protos.ContainerInfo.newBuilder()
+						.setType(Protos.ContainerInfo.Type.DOCKER)
+						.setDocker(Protos.ContainerInfo.DockerInfo.newBuilder()
+							.setNetwork(Protos.ContainerInfo.DockerInfo.Network.HOST)
+							.setForcePullImage(true)
```
Comment (EronWright) [inline flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/MesosApplicationMasterRunner.java]: "setForcePullImage" is probably not appropriate in production; for dev/test you probably need to use an explicit image tag, not 'latest'.

File: flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/MesosApplicationMasterRunner.java [hunk 3]
```diff
@@ -601,6 +601,54 @@ else if (recoveryMode == HighAvailabilityMode.ZOOKEEPER) {
 
 		info.setCommand(cmd);
 
+		// Set base container for task manager if specified in configs.
+		String taskManagerContainerName = flinkConfig.getString(
```
Comment (EronWright) [inline flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/MesosApplicationMasterRunner.java]: The variable name (and associated comment) is inaccurate; may I suggest something like `tmImageName`.

File: flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/MesosApplicationMasterRunner.java (hunk 2 above)
Comment (Makman2) [inline flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/MesosApplicationMasterRunner.java]: Then you can specify the image name with your tag: `myimage:devel`
But we can make a setting for it which defaults to true (I think force-pulling should be on by default, as it has no side-effects).

**Comment by Makman2** [comment]
Btw shall I use "fixup commits" or fix the commit directly? I prefer the second method, I'm not a real fan of non-atomic commits and commits like "Oops, fixing mistake from before". Though I've read inside the contributions-guidelines that the first method is preferred^^

**Comment by mxm** [comment]
It is easier for the review process when you push fixups commits to pull requests. Usually we fixup the commits when merging to master.

**Comment by mxm** [comment]
I'll merge the changes once the comments have been addressed.

File: flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/MesosApplicationMasterRunner.java (hunk 2 above)
Comment (EronWright) [inline flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/MesosApplicationMasterRunner.java]: Force pulling does have side-effects, it causes docker to make additional network requests that might not be desirable.   This behavior is not typical of other DCOS packages, nor is it consistent between the containerizers.   Please turn it off.

**Comment by EronWright** [comment]
@mxm LGTM

**Pull request closed without merging**